### PR TITLE
[Refactor] Refactor dataflow and sync the latest mmengine.

### PR DESCRIPTION
**Pull request by pixeli99**

### CODE REVIEW

File: mmtrack/testing/_utils.py [hunk 1]
```diff
@@ -186,7 +187,7 @@ def demo_mm_inputs(batch_size=1,
                 255,
                 size=(num_key_imgs, *image_shape_group[0]),
                 dtype=np.uint8)
-            mm_inputs['inputs']['img'] = torch.from_numpy(key_img)
+            mm_inputs['inputs']['img'] = torch.from_numpy(key_img).float()
```
Comment (GT9505) [inline mmtrack/testing/_utils.py:190]: delete the `.float()`

File: mmtrack/testing/_utils.py [hunk 2]
```diff
@@ -195,7 +196,7 @@ def demo_mm_inputs(batch_size=1,
                 size=(num_ref_imgs, *image_shape_group[index]),
                 dtype=np.uint8)
             mm_inputs['inputs'][f'{ref_prefix}_img'] = torch.from_numpy(
-                ref_img)
+                ref_img).float()
```
Comment (GT9505) [inline mmtrack/testing/_utils.py:199]: delete the `.float()`

File: tests/test_engine/test_hooks/test_visualization_hook.py [hunk 3]
```diff
@@ -18,23 +18,22 @@ class TestVisualizationHook(TestCase):
...
     def setUp(self) -> None:
         TrackLocalVisualizer.get_instance('visualizer')
-
-        data_sample = TrackDataSample()
-        data_sample.set_metainfo(
-            dict(
-                img_path=osp.join(
-                    osp.dirname(__file__), '../../data/image_1.jpg')))
-        self.data_batch = [dict(data_sample=data_sample)]
+        self.data_batch = [dict(data_sample=None)]
```
Comment (GT9505) [inline tests/test_engine/test_hooks/test_visualization_hook.py:21]: the type of data_batch should be `dict`

File: mmtrack/engine/hooks/visualization_hook.py [hunk 4]
```diff
@@ -92,15 +92,14 @@ def after_val_iter(self, runner: Runner, batch_idx: int,
         total_curr_iter = runner.iter + batch_idx
```
Comment (pixeli99) [inline mmtrack/engine/hooks/visualization_hook.py:92]: fix `batch_idx`

File: tests/test_models/test_vid/test_dff.py [hunk 5]
```diff
@@ -76,9 +76,10 @@ def test_dff_forward_predict_mode(self, cfg_file, devices):
                 for i in range(3):
                     packed_inputs = demo_mm_inputs(
                         batch_size=1, frame_id=i, num_ref_imgs=0)
-                    batch_inputs, data_samples = model.data_preprocessor(
-                        packed_inputs, False)
+                    out_data = model.data_preprocessor(packed_inputs, True)
```
Comment (GT9505) [inline tests/test_models/test_vid/test_dff.py:79]:
```suggestion
                    out_data = model.data_preprocessor(packed_inputs, False)
```
please change it for all unittest of model

File: tests/test_models/test_vid/test_dff.py [hunk 6]
```diff
@@ -45,11 +45,11 @@ def test_dff_forward_loss_mode(self, cfg_file, devices):
...
             packed_inputs = demo_mm_inputs(
                 batch_size=1, frame_id=0, num_ref_imgs=1)
-            batch_inputs, data_samples = model.data_preprocessor(
-                packed_inputs, True)
+            out_data = model.data_preprocessor(packed_inputs, True)
+            inputs, data_samples = out_data['inputs'], out_data['data_samples']
 
             # forward in ``loss`` mode
-            losses = model.forward(batch_inputs, data_samples, mode='loss')
+            losses = model.forward(inputs, data_samples, mode='loss')
```
Comment (GT9505) [inline tests/test_models/test_vid/test_dff.py:52]: better to change to:
```python
            out_data = model.data_preprocessor(packed_inputs, True)
            # forward in ``loss`` mode
            losses = model.forward(**out_data, mode='loss')
```

**Comment by GT9505** [comment]
not forget to change metrics

File: mmtrack/models/data_preprocessors/data_preprocessor.py [hunk 7]
```diff
@@ -118,7 +118,8 @@ def forward(self, data: dict, training: bool = False) -> Dict:
                 imgs = [_img[:, [2, 1, 0], ...] for _img in imgs]
             if self._enable_normalize:
                 imgs = [(_img - self.mean) / self.std for _img in imgs]
-
+            # change to `float`
+            imgs = [_img.float() for _img in imgs]
```
Comment (GT9505) [inline mmtrack/models/data_preprocessors/data_preprocessor.py:122]: better to put this code in the front of  line 119 `if self._enable_normalize:`, in order to be consistent with mmengine

File: tests/test_evaluation/test_metrics/test_sot_metrics.py [hunk 8]
```diff
@@ -47,36 +47,38 @@ def _create_eval_inputs(self, pred_file, gt_file):
...
         self.sot_metric.metrics = ['OPE']
-        gts, preds = self._create_eval_inputs('track_results.txt',
-                                              'gt_for_eval.txt')
-        for gt, pred in zip(gts, preds):
-            self.sot_metric.process([gt], [pred])
+        data_samples = self._create_eval_inputs('track_results.txt',
+                                                'gt_for_eval.txt')
+        for data_sample in data_samples:
+            data_batch = dict(inputs=None, data_samples=None)
+            self.sot_metric.process([data_batch], [data_sample])
```
Comment (GT9505) [inline tests/test_evaluation/test_metrics/test_sot_metrics.py:71]:
```suggestion
            self.sot_metric.process(data_batch, [data_sample])
```

File: tests/test_evaluation/test_metrics/test_sot_metrics.py [hunk 9]
```diff
@@ -47,36 +47,38 @@ def _create_eval_inputs(self, pred_file, gt_file):
...
         assert eval_results[f'{metric_prefix}/norm_precision'] == 70.0
         assert eval_results[f'{metric_prefix}/precision'] == 50.0
 
         # 2. Format-only
         self.sot_metric.format_only = True
-        for gt, pred in zip(gts, preds):
-            self.sot_metric.process([gt], [pred])
+        for data_sample in data_samples:
+            data_batch = dict(inputs=None, data_samples=None)
+            self.sot_metric.process([data_batch], [data_sample])
```
Comment (GT9505) [inline tests/test_evaluation/test_metrics/test_sot_metrics.py:81]:
```suggestion
            self.sot_metric.process(data_batch, [data_sample])
```

File: tests/test_evaluation/test_metrics/test_sot_metrics.py [hunk 10]
```diff
@@ -85,10 +87,11 @@ def test_evaluate(self):
...
         self.sot_metric.metric_options['interval'] = [1, 3]
-        gts, preds = self._create_eval_inputs('vot2018_track_results.txt',
-                                              'vot2018_gt_for_eval.txt')
-        for gt, pred in zip(gts, preds):
-            self.sot_metric.process([gt], [pred])
+        data_samples = self._create_eval_inputs('vot2018_track_results.txt',
+                                                'vot2018_gt_for_eval.txt')
+        for data_sample in data_samples:
+            data_batch = dict(inputs=None, data_samples=None)
+            self.sot_metric.process([data_batch], [data_sample])
```
Comment (GT9505) [inline tests/test_evaluation/test_metrics/test_sot_metrics.py:94]:
```suggestion
            self.sot_metric.process(data_batch, [data_sample])
```

**Review by GT9505 — COMMENTED** [review]
please fix the ut

**Pull request merged**